File: Sorts/java/MergeSort.java
/* MergeSort.java */
public class MergeSort {
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }

    private static void merge(int[] arr, int left, int mid, int right) {
        int size_cpy = right - left + 1;
        int idx_left = 0, idx_right, idx_origin = left;
        int len_left = (mid - left) + 1, len_right = (right - mid) + 1;
        int[] arrCpy = new int[size_cpy];

        System.arraycopy(arr, left, arrCpy, 0, size_cpy);

        idx_right = len_left;
        while (idx_left < len_left && idx_right < size_cpy)
            arr[idx_origin++] =
                (arrCpy[idx_left] <= arrCpy[idx_right] ?
                arrCpy[idx_left++] : arrCpy[idx_right++]);

        while (idx_left < len_left)
            arr[idx_origin++] = arrCpy[idx_left++];
        while (idx_right < size_cpy)
            arr[idx_origin++] = arrCpy[idx_right++];
    }
}

File: Sorts/Python3/MergeSort.py
# MergeSort.py
def MergeSort( arr, left, right ):
	if left >= right: return
	else:
		mid = ( left + right ) // 2
		MergeSort( arr, left, mid )
		MergeSort( arr, mid + 1, right )
		Merge( arr, left, mid, right )

def Merge( arr, left, mid, right ):
	leftCopy = arr[ left : mid + 1 ]
	rightCopy = arr[ mid + 1 : right + 1 ]

	lenLeft, lenRight = len( leftCopy ), len( rightCopy )
	idxLeft, idxRight, idxOrigin = 0, 0, left

	while ( ( idxLeft < lenLeft ) and ( idxRight < lenRight ) ):
		leftVal, rightVal = leftCopy[idxLeft], rightCopy[idxRight]

		if leftVal < rightVal:
			arr[idxOrigin] = leftVal
			idxLeft += 1
		else:
			arr[idxOrigin] = rightVal
			idxRight += 1

		idxOrigin += 1

	while ( idxLeft < lenLeft ):
		arr[idxOrigin] = leftCopy[idxLeft]
		idxOrigin += 1; idxLeft += 1

	while ( idxRight < lenRight ):
		arr[idxOrigin] = rightCopy[idxRight]
		idxOrigin += 1; idxRight += 1
